orchagent/twamporch.cpp

```diff
@@ -0,0 +1,1083 @@
+}
+
+bool TwampOrch::validateUdpPort(uint16_t udp_port)
+{
+    if (udp_port == 862)
+    {
+        return true;
+    }
+    if (udp_port == 863)
+    {
+        return true;
+    }
+    if (udp_port >= 1025)
+    {
+        return true;
+    }
+    return false;
+}
+
+void TwampOrch::increaseTwampSessionCount(void)
+{
+    m_twampSessionCount++;
+}
+
+void TwampOrch::decreaseTwampSessionCount(void)
+{
+    m_twampSessionCount--;
+}
+
+bool TwampOrch::checkTwampSessionCount(void)
+{
+    return m_twampSessionCount < m_maxTwampSessionCount;
+}
+
+void TwampOrch::setSessionStatus(const string& name, const string& status)
+{
+    SWSS_LOG_ENTER();
+
+    vector<FieldValueTuple> fvVector;
+    fvVector.emplace_back(TWAMP_SESSION_STATUS, status);
+    m_stateDbTwampTable.set(name, fvVector);
+}
+
+bool TwampOrch::getSessionStatus(const string &name, string& status)
+{
+    SWSS_LOG_ENTER();
+
+    if (m_stateDbTwampTable.hget(name, TWAMP_SESSION_STATUS, status))
+    {
+        return true;
+    }
+    return false;
+}
+
+void TwampOrch::removeSessionStatus(const string& name)
+{
+	SWSS_LOG_ENTER();
+
+	m_stateDbTwampTable.del(name);
+}
+
+void TwampOrch::removeSessionCounter(const sai_object_id_t session_id)
+{
+	SWSS_LOG_ENTER();
+
+    string key_pattern = "COUNTERS:" + sai_serialize_object_id(session_id) + "*";
+    auto keys = m_countersDb->keys(key_pattern);
+    for (auto& k : keys)
+    {
+        m_countersDb->del(k);
+    }
+}
+
+void TwampOrch::initSessionStats(const string& name)
+{
+    SWSS_LOG_ENTER();
+
+    auto it = m_twampStatistics.find(name);
+    if (it == m_twampStatistics.end())
+    {
+        SWSS_LOG_ERROR("Failed to init non-existent twamp session %s stat info", name.c_str());
+        return;
+    }
+
+    TwampStats& total_stats = it->second;
+
+    total_stats.rx_packets        = 0;
+    total_stats.rx_bytes          = 0;
+    total_stats.tx_packets        = 0;
+    total_stats.tx_bytes          = 0;
+    total_stats.drop_packets      = 0;
+    total_stats.max_latency       = 0;
+    total_stats.min_latency       = 0;
+    total_stats.avg_latency       = 0;
+    total_stats.max_jitter        = 0;
+    total_stats.min_jitter        = 0;
+    total_stats.avg_jitter        = 0;
+    total_stats.avg_latency_total = 0;
+    total_stats.avg_jitter_total  = 0;
+}
+
+bool TwampOrch::registerTwampEventNotification(void)
+{
+    sai_attribute_t attr;
+    sai_status_t status;
+    sai_attr_capability_t capability;
+
+    status = sai_query_attribute_capability(gSwitchId, SAI_OBJECT_TYPE_SWITCH,
+                                            SAI_SWITCH_ATTR_TWAMP_SESSION_EVENT_NOTIFY,
+                                            &capability);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Unable to query the TWAMP event notification capability");
+        return false;
+    }
+
+    if (!capability.set_implemented)
+    {
+        SWSS_LOG_ERROR("TWAMP register event notification not supported");
```

Comment:
IMO, this is not an error but a notice as SAI may not implement this feature. Please change

Reply:
Done